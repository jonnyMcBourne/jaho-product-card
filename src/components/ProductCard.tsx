import React from "react";
import { createContext, CSSProperties } from "react";
import styles from "../styles/styles.module.css";
import { useProduct } from "../hooks/useProduct";
import { ProductContextProps, Product, onChangeArgs,InitialValues, ProductCardHandle } from "../interfaces/productInterfaces";

export interface Props {
  product: Product;
  //children?: ReactElement | ReactElement[];
  children: (args:ProductCardHandle) => JSX.Element;
  className?: string;
  style?: CSSProperties;
  onChange?:(arg:onChangeArgs)=>void;
  value?:number;
  initialValues?: InitialValues;
}

export const ProductContext = createContext({} as ProductContextProps);

const { Provider } = ProductContext;

export const ProductCard = ({
  children,
  product,
  className,
  style,
  onChange,
  value,
  initialValues,
}: Props) => {
  const { counter, increaseBy, isMaxCountReached, maxCount, resetCounter } =
    useProduct({ product, onChange, value, initialValues });

  return (
    <Provider value={{ counter, increaseBy, product, maxCount }}>
      <div className={`${styles.productCard} ${className}`} style={style}>
        {children({
          count: counter,
          increaseBy: increaseBy,
          isMaxCountReached: isMaxCountReached,
          maxCount: initialValues?.maxCount,
          product: product,
          reset: resetCounter,
        })}
      </div>
    </Provider>
  );
};
export default ProductCard;


