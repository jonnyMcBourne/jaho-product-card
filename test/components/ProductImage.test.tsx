import React from "react";
import renderer from 'react-test-renderer';
import { ProductImage, ProductCard } from '../../src/components';
import { products } from '../data/products';

describe('productImage',()=>{
    test('must show the component without an image',()=>{
        const wrapper = renderer.create(<ProductImage img="https://th.bing.com/th/id/OIP.UPn4LXlCa5zuvmtt_rhlUgHaHa?pid=ImgDet&rs=1"/>)
        expect(wrapper.toJSON()).toMatchSnapshot();
    });
});

describe('productImage', () => {
  test('must show image of the product inside the productCard', () => {
    const wrapper = renderer.create(
      <ProductCard product={products[1]}>
        {({}) => (
          <>
            <ProductImage />
          </>
        )}
      </ProductCard>
    );
    expect(wrapper.toJSON()).toMatchSnapshot();
  });
});