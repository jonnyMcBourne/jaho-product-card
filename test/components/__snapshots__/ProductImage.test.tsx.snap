// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`productImage must show image of the product inside the productCard 1`] = `
<div
  className="productCard undefined"
>
  <img
    alt="no-name-provided"
    className="productImg undefined"
    src="https://th.bing.com/th/id/OIP.UPn4LXlCa5zuvmtt_rhlUgHaHa?pid=ImgDet&rs=1"
  />
</div>
`;

exports[`productImage must show the component without an image 1`] = `
<img
  className="productImg undefined"
  src="https://th.bing.com/th/id/OIP.UPn4LXlCa5zuvmtt_rhlUgHaHa?pid=ImgDet&rs=1"
/>
`;
